File: atm-cockpit/src/shared/sdui/validate.mjs
// Minimal JSON-Schema validator — zero dependencies, runnable under plain `node`.
// Supports exactly the keywords our schemas use (the same fail-closed spirit as
// the brain's hand-rolled validator). Returns an array of error strings; empty = valid.

const TYPE = {
  object: (v) => v !== null && typeof v === "object" && !Array.isArray(v),
  array: (v) => Array.isArray(v),
  string: (v) => typeof v === "string",
  integer: (v) => typeof v === "number" && Number.isInteger(v),
  number: (v) => typeof v === "number",
  boolean: (v) => typeof v === "boolean",
  null: (v) => v === null,
};

export function validate(instance, schema, path = "$") {
  const errors = [];
  walk(instance, schema, path, errors);
  return errors;
}

function walk(inst, schema, path, errors) {
  if (schema.type !== undefined) {
    const types = Array.isArray(schema.type) ? schema.type : [schema.type];
    if (!types.some((t) => (TYPE[t] || (() => true))(inst))) {
      errors.push(`${path}: expected type ${schema.type}, got ${typeName(inst)}`);
      return; // dependent checks assume the type held
    }
  }
  if ("const" in schema && !deepEqual(inst, schema.const)) {
    errors.push(`${path}: must equal ${JSON.stringify(schema.const)}`);
  }
  if (schema.enum && !schema.enum.some((e) => deepEqual(e, inst))) {
    errors.push(`${path}: ${JSON.stringify(inst)} not in ${JSON.stringify(schema.enum)}`);
  }
  if (typeof inst === "string") {
    if (schema.pattern && !new RegExp(schema.pattern).test(inst)) {
      errors.push(`${path}: ${JSON.stringify(inst)} does not match /${schema.pattern}/`);
    }
    if (schema.maxLength !== undefined && inst.length > schema.maxLength) {
      errors.push(`${path}: longer than maxLength ${schema.maxLength}`);
    }
    if (schema.minLength !== undefined && inst.length < schema.minLength) {
      errors.push(`${path}: shorter than minLength ${schema.minLength}`);
    }
  }
  if (typeof inst === "number") {
    if (schema.minimum !== undefined && inst < schema.minimum) {
      errors.push(`${path}: ${inst} below minimum ${schema.minimum}`);
    }
    if (schema.maximum !== undefined && inst > schema.maximum) {
      errors.push(`${path}: ${inst} above maximum ${schema.maximum}`);
    }
  }
  if (TYPE.object(inst)) {
    for (const req of schema.required || []) {
      if (!(req in inst)) errors.push(`${path}: missing required field '${req}'`);
    }
    const props = schema.properties || {};
    if (schema.additionalProperties === false) {
      for (const key of Object.keys(inst)) {
        if (!(key in props)) errors.push(`${path}: unexpected property '${key}'`);
      }
    }
    for (const [key, sub] of Object.entries(props)) {
      if (key in inst) walk(inst[key], sub, `${path}.${key}`, errors);
    }
  }
  if (Array.isArray(inst)) {
    if (schema.minItems !== undefined && inst.length < schema.minItems) {
      errors.push(`${path}: fewer than minItems ${schema.minItems}`);
    }
    if (schema.maxItems !== undefined && inst.length > schema.maxItems) {
      errors.push(`${path}: more than maxItems ${schema.maxItems}`);
    }
    if (schema.uniqueItems) {
      const seen = [];
      for (const it of inst) {
        if (seen.some((s) => deepEqual(s, it))) { errors.push(`${path}: items must be unique`); break; }
        seen.push(it);
      }
    }
    if (schema.items && typeof schema.items === "object") {
      inst.forEach((it, i) => walk(it, schema.items, `${path}[${i}]`, errors));
    }
  }
}

function typeName(v) {
  if (v === null) return "null";
  if (Array.isArray(v)) return "array";
  return typeof v;
}

function deepEqual(a, b) {
  if (a === b) return true;
  if (typeof a !== typeof b) return false;
  if (a && b && typeof a === "object") {
    const ka = Object.keys(a), kb = Object.keys(b);
    if (Array.isArray(a) !== Array.isArray(b)) return false;
    if (ka.length !== kb.length) return false;
    return ka.every((k) => deepEqual(a[k], b[k]));
  }
  return false;
}

File: Atm_second-brain/server/validate.py
"""Minimal JSON-Schema validator for the note contract. Standard library only.

Supports exactly the keywords schema.vN.json uses: type, required, const, enum,
pattern, properties, items, minLength, minimum, uniqueItems, additionalProperties
(only the boolean form is honored loosely). Returns a list of human-readable
errors; empty list means valid.

This is deliberately not a general validator — it's the enforcement arm of
write_with_provenance, kept small enough to audit by eye.
"""
from __future__ import annotations

import json
import os
import re
from functools import lru_cache
from typing import Any

import config

_TYPE_CHECKS = {
    "object": lambda v: isinstance(v, dict),
    "array": lambda v: isinstance(v, list),
    "string": lambda v: isinstance(v, str),
    "integer": lambda v: isinstance(v, int) and not isinstance(v, bool),
    "number": lambda v: isinstance(v, (int, float)) and not isinstance(v, bool),
    "boolean": lambda v: isinstance(v, bool),
    "null": lambda v: v is None,
}


@lru_cache(maxsize=4)
def load_schema(version: int) -> dict:
    path = os.path.join(config.SCHEMA_DIR, f"schema.v{version}.json")
    with open(path, encoding="utf-8") as fh:
        return json.load(fh)


def validate(instance: Any, schema: dict, path: str = "") -> list[str]:
    errors: list[str] = []
    _validate(instance, schema, path or "<root>", errors)
    return errors


def _validate(inst: Any, schema: dict, path: str, errors: list[str]) -> None:
    t = schema.get("type")
    if t is not None:
        types = t if isinstance(t, list) else [t]
        if not any(_TYPE_CHECKS.get(tc, lambda _v: True)(inst) for tc in types):
            errors.append(f"{path}: expected type {t}, got {type(inst).__name__}")
            return  # further checks assume the type held

    if "const" in schema and inst != schema["const"]:
        errors.append(f"{path}: must equal {schema['const']!r}, got {inst!r}")

    if "enum" in schema and inst not in schema["enum"]:
        errors.append(f"{path}: {inst!r} not in allowed {schema['enum']}")

    if isinstance(inst, str):
        pat = schema.get("pattern")
        if pat and not re.search(pat, inst):
            errors.append(f"{path}: {inst!r} does not match pattern {pat!r}")
        if "minLength" in schema and len(inst) < schema["minLength"]:
            errors.append(f"{path}: shorter than minLength {schema['minLength']}")

    if isinstance(inst, (int, float)) and not isinstance(inst, bool):
        if "minimum" in schema and inst < schema["minimum"]:
            errors.append(f"{path}: {inst} below minimum {schema['minimum']}")

    if isinstance(inst, dict):
        for req in schema.get("required", []):
            if req not in inst:
                errors.append(f"{path}: missing required field '{req}'")
        props = schema.get("properties", {})
        for key, subschema in props.items():
            if key in inst:
                _validate(inst[key], subschema, f"{path}.{key}", errors)

    if isinstance(inst, list):
        if schema.get("uniqueItems"):
            seen, dupes = [], False
            for item in inst:
                if item in seen:
                    dupes = True
                    break
                seen.append(item)
            if dupes:
                errors.append(f"{path}: items must be unique")
        item_schema = schema.get("items")
        if isinstance(item_schema, dict):
            for i, item in enumerate(inst):
                _validate(item, item_schema, f"{path}[{i}]", errors)


def validate_frontmatter(frontmatter: dict, version: int) -> list[str]:
    return validate(frontmatter, load_schema(version))
